File: src/main/java/com/ultreon/mods/screenshotmanager/common/Resizer.java
package com.ultreon.mods.screenshotmanager.common;

import lombok.Getter;

public class Resizer {
    @Getter
    private final float ratio;
    @Getter
    private final float relativeRatio;
    @Getter
    private final Orientation orientation;
    @Getter
    private final float sourceWidth;
    @Getter
    private final float sourceHeight;

    public Resizer(float srcWidth, float srcHeight) {
        this.ratio = srcWidth / srcHeight;

        if (srcWidth > srcHeight) {
            this.relativeRatio = srcWidth / srcHeight;
            this.orientation = Orientation.LANDSCAPE;
        } else if (srcWidth < srcHeight) {
            this.relativeRatio = srcHeight / srcWidth;
            this.orientation = Orientation.PORTRAIT;
        } else {
            this.relativeRatio = 1;
            this.orientation = Orientation.SQUARE;
        }

        this.sourceWidth = srcWidth;
        this.sourceHeight = srcHeight;
    }

    public FloatSize thumbnail(float maxWidth, float maxHeight) {
        float aspectRatio;
        float width;
        float height;

        if (sourceWidth > sourceHeight) {
            aspectRatio = (float) (sourceWidth / (double) sourceHeight);

            width = maxWidth;
            height = (int) (width / aspectRatio);

            if (height > maxHeight) {
                aspectRatio = (float) (sourceHeight / (double) sourceWidth);

                height = maxHeight;
                width = (int) (height / aspectRatio);
            }
        } else {
            aspectRatio = (float) (sourceHeight / (double) sourceWidth);

            height = maxHeight;
            width = (int) (height / aspectRatio);
            if (width > maxWidth) {
                aspectRatio = (float) (sourceWidth / (double) sourceHeight);

                width = maxWidth;
                height = (int) (width / aspectRatio);
            }
        }

        return new FloatSize(width, height);
    }

    /**
     * Aspect ratio orientation.
     */
    public enum Orientation {
        LANDSCAPE,
        SQUARE,
        PORTRAIT
    }
}

File: common/src/main/java/com/ultreon/mods/screenshotmanager/util/Resizer.kt
package com.ultreon.mods.screenshotmanager.util

class Resizer(srcWidth: Float, srcHeight: Float) {
    val ratio: Float
    var relativeRatio = 0f
    var orientation: Orientation? = null
    val sourceWidth: Float
    val sourceHeight: Float

    init {
        ratio = srcWidth / srcHeight
        if (srcWidth > srcHeight) {
            relativeRatio = srcWidth / srcHeight
            orientation = Orientation.LANDSCAPE
        } else if (srcWidth < srcHeight) {
            relativeRatio = srcHeight / srcWidth
            orientation = Orientation.PORTRAIT
        } else {
            relativeRatio = 1f
            orientation = Orientation.SQUARE
        }
        sourceWidth = srcWidth
        sourceHeight = srcHeight
    }

    fun thumbnail(maxWidth: Float, maxHeight: Float): FloatSize {
        var aspectRatio: Float
        var width: Float
        var height: Float
        if (sourceWidth > sourceHeight) {
            aspectRatio = (sourceWidth / sourceHeight.toDouble()).toFloat()
            width = maxWidth
            height = (width / aspectRatio).toInt().toFloat()
            if (height > maxHeight) {
                aspectRatio = (sourceHeight / sourceWidth.toDouble()).toFloat()
                height = maxHeight
                width = (height / aspectRatio).toInt().toFloat()
            }
        } else {
            aspectRatio = (sourceHeight / sourceWidth.toDouble()).toFloat()
            height = maxHeight
            width = (height / aspectRatio).toInt().toFloat()
            if (width > maxWidth) {
                aspectRatio = (sourceWidth / sourceHeight.toDouble()).toFloat()
                width = maxWidth
                height = (width / aspectRatio).toInt().toFloat()
            }
        }
        return FloatSize(width, height)
    }

    /**
     * Aspect ratio orientation.
     */
    enum class Orientation {
        LANDSCAPE, SQUARE, PORTRAIT
    }
}